k2/csrc/ragged_ops.cu

```diff
@@ -560,6 +601,157 @@ static RaggedShape IndexAxis0(RaggedShape &src, const Array1<int32_t> &new2old,
+                *new_offsets_data = new_offsets.Data();
+
+  if (num_axes == 2) {
+    if (c->GetDeviceType() == kCpu) {
+    } else {
+      K2_CHECK_EQ(c->GetDeviceType(), kCuda);
+      int32_t ans_num_elems = new_offsets.Back();
+      Array1<int32_t> ans_row_splits(c, ans_num_elems);
+      int32_t *ans_row_splits_data = ans_row_splits.Data();
+      if (elem_indexes != nullptr)
+        *elem_indexes = Array1<int32_t>(c, ans_num_elems);
+      int32_t *elem_indexes_data =
+          elem_indexes != nullptr ? elem_indexes->Data() : nullptr;
+      if (ans_num_elems == 0) return RaggedShape2(&new_offsets, nullptr, 0);
+
+      mgpu::context_t *mgpu_context = GetModernGpuAllocator(c);
+      auto lambda_set_ans = [=] __device__(int32_t index, int32_t seg,
+                                           int32_t rank) {
+        ans_row_splits_data[new_offsets_data[seg] + rank] = seg;
+        if (elem_indexes_data != nullptr) {
+          elem_indexes_data[new_offsets_data[seg] + rank] =
+              old_offsets_data[seg] + rank;
+        }
+      };
+      K2_CUDA_SAFE_CALL(mgpu::transform_lbs(
+          lambda_set_ans, new_offsets.Back(), new_offsets.Data(),
+          new_offsets.Dim() - 1, *mgpu_context));
+      RaggedShape ans =
+          RaggedShape2(&new_offsets, &ans_row_splits, ans_num_elems);
+      return ans;
+    }
+  }
+
+  Array1<int32_t> tot_sizes_out(c, num_axes);
+  int32_t *tot_sizes_out_data = tot_sizes_out.Data();
+  K2_EVAL(
+      c, num_axes, lambda_set_tot_sizes, (int32_t i)->void {
+        if (i == 0)
+          tot_sizes_out_data[0] = ans_dim0;
+        else {
+          tot_sizes_out_data[i] = new_offsets_data[ans_dim0 * i] -
+                                  new_offsets_data[ans_dim0 * (i - 1)];
+        }
+      });
+  tot_sizes_out = tot_sizes_out.To(GetCpuContext());
+
+  if (elem_indexes != nullptr)
+    *elem_indexes = Array1<int32_t>(c, tot_sizes_out.Back());
+  int32_t *elem_indexes_data =
+      elem_indexes != nullptr ? elem_indexes->Data() : nullptr;
+  RaggedShape ans = RaggedShapeFromTotSizes(c, num_axes, tot_sizes_out.Data());
+
+  RowSplitsAccessor<5> ans_row_splits_acc(ans);
+  RowIdsAccessor<5> ans_row_ids_acc(ans);
+
+  RowSplitsAccessor<5> src_row_splits_acc(src);
+  RowIdsAccessor<5> src_row_ids_acc(src);
+
+  ans.Layers()[0].row_splits = new_offsets.Arange(0, ans_dim0 + 1);
+
+  if (c->GetDeviceType() == kCpu) {
+  } else {
+    K2_CHECK_EQ(c->GetDeviceType(), kCuda);
+
+    int32_t tot_elems = new_offsets.Back();
+    if (tot_elems == 0) {
+      for (int32_t i = 0; i != num_axes - 1; ++i) {
+        if (i != 0)
+          ans.Layers()[i].row_splits =
+              new_offsets.Arange(ans_dim0 + i, ans_dim0 + i + 1);
+        ans.Layers()[i].row_ids = Array1<int32_t>(c, 0);
+      }
+      return ans;
+    }
+    mgpu::context_t *mgpu_context = GetModernGpuAllocator(c);
+    auto lambda_set_ans = [=] __device__(int32_t index, int32_t seg,
```

Comment:
Would be nice to have some documentation of the args of this lambda.
The documntation of transform_lbs
https://moderngpu.github.io/doc/api.html
unfortunately doesn't seem to explain this.

Reply:
.. I know this is not for merge, but some explanation would be nice.

Reply:
Sure, mainly 

`index` is the index of element (i.e. idx01)
`seg` is the row id (i.e. idx0),
`rank` is the index in current `seg`/row (i.e. idx1)

also added the documentation in the code, thanks.